File: python/flesch.py
#!/usr/bin/python3
#Emily Wilbourn
#Python flesch program


#importing sys so that I can receive command line input
import sys

#function to fill a set with all the words from a text file amd return number of sentences
#precondition: pass in the file name (string) and the set we want to update
#postcondition: returns the number of sentences in a word and updates our set
def tokenizeFile(words, fileName):
	numSentences = 0
	with open(fileName,encoding='utf8',errors='ignore') as file: 
		for line in file:
			for word in line.split():
				if (not word.isdigit()):
					for char in word:
						if findPunctuation(char):
							numSentences+=1
							word = word.replace(char,"")
						#remove unwanted characters from the file
						if char in ",[]" or char == "'":
							word = word.replace(char,"")
					words.append(word.lower())
					#print(word.lower())
	return numSentences

	
#function to determine if a word has punctuation
#precondition: pass in the word being stored in a list of strings
#postcondition: return True if there is punctuat
def findPunctuation(character):
	punctuation = ".:;!?"
	if character in punctuation:
		return True
	return False

#function to determine if a character is a vowel
#precondition: pass in a character
#postcondition: return True is the character is a vowel, False if it isn't
def findVowel(character):
	vowels = "aeiouy"
	if character in vowels:
		return True
	return False

#function to count the number of syllables in a word (string)
#precondition: pass in a word (string)
#postcondition: return the number of syllables in a word (integer)
def findSyllables(word):
	syllables = 0
	try:
		if findVowel(word[0]):
			syllables+=1
	except:	
		syllables = 0
	for i in range(1, len(word)):
	#this handles when the word ends in an e (which is silent, and thus not a syllable)
		if(not ((word[i]).lower() == 'e' and i == (len(word)-1))):
			#handles two vowels in a row
			#i.e. spool, cool, moon
			if findVowel(word[i]) and not findVowel(word[i-1]):
				syllables+=1
				
	if syllables == 0:
		syllables = 1

	return syllables

#function to count the number of syllables in the file
#precondition: pass in the list of words (strings)
#postcondition: return the number of syllables in the text file (integer)
def totalSyllables(words):
	numSyllables = 0
	for word in words:
		numSyllables += findSyllables(word)	
	
	return numSyllables

#function to create a list of Dale Chall words
#precondition: pass in nothing
#postcondition: return the set of words (strings)
def createDaleChallList():
	daleChall = []
	with open("/pub/pounds/CSC330/dalechall/wordlist1995.txt",encoding='utf8',errors='ignore') as file:
		for line in file:
			for word in line.split():
				for char in word:
					if char == "'":
						word = word.replace("'","")
				daleChall.append(word.lower())
	daleChall.sort()
	return daleChall

#function to quickly find values in list
#precondition: pass in the list of words (string) and the key we're looking for (string)
#postcondition: return the location of the key in list; if key isn't in list, return -1
def binarySearch(wordList,key):
	first = 0
	last = len(wordList)-1
	found = False
	while (first<=last and not found): 
		mid = (first+last)//2
		if wordList[mid] == key:
			found = True
		else:
			if key < wordList[mid]:
				last = mid - 1
			else:
				first = mid + 1
	return found

#function to count the number of challenging words in a text file based on the Dale Chall List
#precondition: pass in the set of words (strings)
#postcondition: return the number of challenging words (integer)
def countChallengingWords(words):
	difficultWords = 0
	daleChall = []
	daleChall = createDaleChallList()
	daleChall = set(daleChall)
	for word in words:
		if (word not in daleChall):
			difficultWords+=1
	return difficultWords

#function to calculate the flesch-kincaid readability index
#precondition: pass in the set of words (string)
#postcondition: returns the flesch-kincaid readability index (float to 1 decimal place)
def fleschKincaidIndex(words, numSentences):
	numWords = len(words)
	totalSyll = totalSyllables(words)

	#these are floats
	a = totalSyll/numWords
	b = numWords/numSentences

	#index is a float
	index = (a*11.8) + (b*0.39) - 15.59

	#return the index (float) rounded to 1 decimal place
	return round(index, 1)


#function to calculate the flesch readability index
#precondition: pass in the list of words (string)
#postcondition: returns the flesch index (integer)
def fleschIndex(words, numSentences):
        numWords = len(words)
        totalSyll = totalSyllables(words)

	#these are floats
        a = totalSyll/numWords
        b = numWords/numSentences

	#return the index (float) rounded to 1 decimal place
        return round((206.835 - (a*84.6) - (b*1.015)))

#function to calculate the Dale-Chall Index
#precondition: pass in the list of words (string)
#postcondition: returns the Dale-Chall Index (float rounded to 1 decimal place)
def daleChallIndex(words, numSentences):
	numWords = len(words)
	difficultWords =  countChallengingWords(words)

	#these are floats
	a = difficultWords/numWords
	b = numWords/numSentences
	index = ((a*100)*0.1579)+(b*0.0496)
	if((a*100) > 5):
		index += 3.6365;

	#return the index (float) rounded to 1 decimal place
	return round(index, 1);



#####################################################################################

def main():
	#receives command line input for the file name
	#command to enter for KJV.txt: python3 flesch.py KJV
	#here, KJV will be stored in inputF
	inputF = str(sys.argv[1])

	
	inputFile = "/pub/pounds/CSC330/translations/" + inputF + ".txt"

	#instantiate a list
	words = []

	#fill up the list with the words from the text file
	numSentences = tokenizeFile(words,inputFile)
	numWords = len(words)
	numSyllables = totalSyllables(words)
	numChallWords = countChallengingWords(words)
	daleChall = daleChallIndex(words, numSentences)
	flesch = fleschIndex(words, numSentences)
	fleschKincaid = fleschKincaidIndex(words, numSentences)
	
	if(len(inputF) == 3):
		print("Python    " , inputF, "         " , flesch, "     ", fleschKincaid, "            ", daleChall,"          ");

	if(len(inputF) == 4):
		print("Python    " , inputF, "        " , flesch, "     ", fleschKincaid, "            ", daleChall,"          ");
	if(len(inputF) == 5):
		print("Python    " , inputF, "       " , flesch, "     ", fleschKincaid, "            ", daleChall,"          ");
#################################################################################

main()

File: perl/flesch.pl
#!/usr/bin/perl
#Emily Wilbourn
#Perl flesch program

use strict;
use warnings;

#using this to determine if a character is a number or not when 
#reading in input from the text file
use Scalar::Util qw(looks_like_number);


# Grab the name of the file from the command line, exit if no name given
#just take in the name of the file, not the whole path
#i.e. perl flesch.pl KJV
my $name = $ARGV[0] or die "Need to get file name on the command line\n";


#variable to hold the full file path
my $filename = '/pub/pounds/CSC330/translations/' . $name . '.txt';

# Use the filename
open(DATA, "<$filename") or die "Couldn't open file $filename, $!";


# Open the Dale-Chall list and collect the data inside it
open(DATA1, "<", "/pub/pounds/CSC330/dalechall/wordlist1995.txt") or die "Couldn't open file $filename, $!";

#create an array that holds all the words in the dale-chall list
my @dale_chall = <DATA1>;

#The next line puts all the lines from the text file into an array called @all_lines
my @all_lines = <DATA>;

close DATA;
#array to hold individual words
my @all_words = ();

#create a variable that will count the words
my $word_count = 0;

#create a variable that will count the sentences
my $sentence_count = 0;

#instantiate an array that holds all the punctuation
my @punctuation = (".", ";", ":", "!", "?");

#array that holds other characters that we need to 
#remove from the words we are reading in
my @other_punctuation = ("[", ",", "]");

#define a vowel list that we will use to find vowels in words
my @vowel_list = ("a","e","i","o","u","y");
my $syllable_count = 0;



#Now take each line and break it into tokens based on spaces and print the token
foreach my $line (@all_lines)
{
	my @tokens = split(' ', $line);
	foreach my $token (@tokens)
	{
		#instantiating a character array where each index is a character
		#in the word we are currently looking at
		my @char_array = split(//, $token);

		#increment the word counter variable
		$word_count++ if(!looks_like_number($token));
		for my $char(@char_array)
		{
			for my $punct (@punctuation)
			{
				$sentence_count++ if(($char cmp $punct) == 0);
				$token =~ s/\Q$char// if (($char cmp $punct) == 0);
				for my $other_punct(@other_punctuation)
				{	
					$token =~ s/\Q$char// if (($char cmp $other_punct) == 0);
				}
			}
			
		}
		#make the token, aka the word, all lowercase
		my $token1 = lc $token;

		#if the token, aka the word, doesn't look like a number,
		#then put it into the array of words
		push(@all_words, $token1) if(!looks_like_number($token));
	}
}

#loop to count the number of syllables in the file
my $total_syll_count = 0;
for my $word (@all_words)
{
	$syllable_count = 0;
	my @chars = split("", $word);
	for my $v (@vowel_list)
	{
		#if the first character in a word is defined, then determine if it 
		#is a vowel or not 
		if (defined $chars[0])
		{
			$syllable_count++ if (($chars[0] cmp $v)==0);
		}        
	}
	for (my $i = 1; $i < $#chars; $i++) 
	{
        	my $char = $chars[$i];
		my $prev_char = $chars[$i-1];
		#this handles when the word ends in an e (which is silent, and thus not a syllable)
        	
		if(not (($char cmp 'e') == 0 and $i == ($#chars-1)) and (defined $char))
		{
        		#handles two vowels in a row
        		#i.e. spool, cool, moon
        		for my $v (@vowel_list)
                        {
                         	       
				$syllable_count++ if(($char cmp $v) == 0 && ($prev_char cmp $v) !=0);
			}
        	}
	}
	if ($syllable_count == 0)
	{
        	$syllable_count = 1
	}
	
	$total_syll_count += $syllable_count;
	
}

#initialize variable to hold the number of difficult words
my $difficult_word_count = 0;
my @new_dale_chall = ();
#loop to fill up an array with the Dale-Chall words
foreach my $line (@dale_chall)
{
        my @tokens = split(' ', $line);
        foreach my $token (@tokens)
        {
		#make the token, aka the word, all lowercase
		my $token1 = lc $token;
		push(@new_dale_chall, $token1);
	}
}

@new_dale_chall = sort{$a cmp $b} @new_dale_chall;
#calculate the number of difficult words in the text file
#by comparing the contents to the dale-chall list by utilizing a 
#binary search
foreach my $key(@all_words)
{

	my $low = 0;
	my $mid;
	my $found_key = 0;
	my $high = $#new_dale_chall;
	my $index;
		
	while( ( $low <= $high ) && !$found_key ) 
	{
  		$mid =int( ( $low + $high ) / 2);
  		if(( $key cmp $new_dale_chall[$mid] ) == 0)
		{
    			$found_key = 1;
    			$index = int( $mid );
  		}

  		elsif(( $key cmp $new_dale_chall[$mid] ) < 0)
		{
    			$high = $mid - 1;
  		}

  		else 
		{
    			$low = $mid + 1;
  		}
	}

	if( $found_key ) 
	{
	}

	else 
	{
		$difficult_word_count++;	
	}
	
}


#calculate flesch readability index
my $a1 = ($total_syll_count)/($word_count);
my $b1 = ($word_count)/($sentence_count);
my $flesch_index = 206.835-(($a1)*84.6)-($b1)*(1.015);

#round the flesch index to the nearest integer
my $round_f = sprintf("%.0f",$flesch_index);

#calculate flesch-kincaid readability index
my $flesch_kincaid = ($a1)*11.8+($b1)*0.39-15.59;

#round the flesch-kincaid index to one decimal place
my $round_fk = sprintf("%.1f", $flesch_kincaid);

#calculate dale-chall index
my $a = ($difficult_word_count)/($word_count);
my $b = ($word_count)/($sentence_count);

my $d_c_index = (($a*100)*0.1579)+(($b)*0.0496);
$d_c_index = $d_c_index+3.6365 if(($a*100) > 5);

#round the dale-chall index to one decimal place
my $round_dc = sprintf("%.1f", $d_c_index);


#print out the output based on the length of the filename we read in on the command line 
print "Perl       " , $name, "           " , $round_f, "       ", $round_fk, "              ", $round_dc,"          " if (length($name) == 3);
print "Perl       " , $name, "          " , $round_f, "       ", $round_fk, "              ", $round_dc,"          " if (length($name) == 4);
print "Perl       " , $name, "         " , $round_f, "       ", $round_fk, "              ", $round_dc,"          " if (length($name) == 5);
print"\n";

